jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java

```diff
@@ -0,0 +1,129 @@
+  public enum LayerType {
+    DEPENDENCIES("dependencies", JavaEntrypointConstructor.DEFAULT_DEPENDENCIES_PATH_ON_IMAGE),
+    SNAPSHOT_DEPENDENCIES(
+        "snapshot dependencies", JavaEntrypointConstructor.DEFAULT_DEPENDENCIES_PATH_ON_IMAGE),
+    RESOURCES("resources", JavaEntrypointConstructor.DEFAULT_RESOURCES_PATH_ON_IMAGE),
+    CLASSES("classes", JavaEntrypointConstructor.DEFAULT_CLASSES_PATH_ON_IMAGE),
+    EXTRA_FILES("extra files", "/");
+
+    private final String label;
+    private final String extractionPath;
+
+    /** Initializes with a label for the layer and the layer files' default extraction path root. */
+    LayerType(String label, String extractionPath) {
+      this.label = label;
+      this.extractionPath = extractionPath;
+    }
+
+    public String getLabel() {
+      return label;
+    }
+
+    public String getExtractionPath() {
+      return extractionPath;
+    }
+  }
+
+  /** Builds with each layer's files. */
+  public static class Builder {
+
+    private Map<LayerType, List<Path>> layerFilesMap = new EnumMap<>(LayerType.class);
+
+    private Builder() {
+      for (LayerType layerType : LayerType.values()) {
+        layerFilesMap.put(layerType, new ArrayList<>());
+      }
+    }
+
+    public Builder setDependenciesFiles(List<Path> dependenciesFiles) {
+      layerFilesMap.put(LayerType.DEPENDENCIES, dependenciesFiles);
+      return this;
+    }
+
+    public Builder setSnapshotDependenciesFiles(List<Path> snapshotDependenciesFiles) {
+      layerFilesMap.put(LayerType.SNAPSHOT_DEPENDENCIES, snapshotDependenciesFiles);
+      return this;
+    }
+
+    public Builder setResourcesFiles(List<Path> resourcesFiles) {
+      layerFilesMap.put(LayerType.RESOURCES, resourcesFiles);
+      return this;
+    }
+
+    public Builder setClassesFiles(List<Path> classesFiles) {
+      layerFilesMap.put(LayerType.CLASSES, classesFiles);
+      return this;
+    }
+
+    public Builder setExtraFiles(List<Path> extraFiles) {
+      layerFilesMap.put(LayerType.EXTRA_FILES, extraFiles);
+      return this;
+    }
+
+    public JavaLayerConfigurations build() {
+      ImmutableMap.Builder<LayerType, LayerConfiguration> layerConfigurationsMap =
+          ImmutableMap.builderWithExpectedSize(LayerType.values().length);
+      for (LayerType layerType : LayerType.values()) {
+        List<Path> layerFiles = Preconditions.checkNotNull(layerFilesMap.get(layerType));
+        layerConfigurationsMap.put(
+            layerType,
+            LayerConfiguration.builder()
+                .addEntry(layerFiles, layerType.getExtractionPath())
+                .setLabel(layerType.getLabel())
+                .build());
+      }
+      return new JavaLayerConfigurations(layerConfigurationsMap.build());
+    }
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  private final ImmutableMap<LayerType, LayerConfiguration> layerConfigurationMap;
```

Comment:
Any reason we can't use the layers directly instead of using a map? I would think our JavaLayerConfiguration is going to be pretty static.

Reply:
If this is to fit in with some future interface, I think we should include the interface in this PR? And maybe even then, could `JavaLayerConfigurations` hide the "map"-ness of the interface and provide stronger accessors?